### PR TITLE
Avoid an extra call to f in Statistic.mean(f, A)

### PR DESCRIPTION
**Pull request by kagalenko-m-b**
Proposed fix by @stevengj for the issue #49

### CODE REVIEW

File: src/Statistics.jl [hunk 1]
```diff
@@ -316,7 +326,7 @@ whereas the sum is scaled with `n` if `corrected` is
 
 If `itr` is an `AbstractArray`, `dims` can be provided to compute the variance
 over dimensions. In that case, `mean` must be an array with the same shape as
-`mean(itr, dims=dims)` (additional trailing singleton dimensions are allowed).
+`mean(itr; dims)` (additional trailing singleton dimensions are allowed).
```
Comment (nalimilan) [inline src/Statistics.jl:329]: Please revert unrelated changes (PR should be focused on a single issue at a time). This one can also be misleading, as if users naively replace `dims` with e.g. `1` the call will fail.

File: test/runtests.jl [hunk 2]
```diff
@@ -162,6 +162,15 @@ end
     @test (@inferred mean(Iterators.filter(x -> true, Int[]))) === 0/0
     @test (@inferred mean(Iterators.filter(x -> true, Float32[]))) === 0.f0/0
     @test (@inferred mean(Iterators.filter(x -> true, Float64[]))) === 0/0
+    # Check that mean does not call function argument an extra time
+    let  _cnt = 0, N = 100, x = rand(Int, N)
```
Comment (nalimilan) [inline test/runtests.jl:166]:
```suggestion
    let _cnt = 0, N = 100, x = rand(Int, N)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -162,6 +162,15 @@ end
...
+        f(x) = begin; _cnt += 1; x; end
```
Comment (nalimilan) [inline test/runtests.jl:167]:
```suggestion
        f(x) = (_cnt += 1; x)
```

File: src/Statistics.jl [hunk 4]
```diff
@@ -165,16 +165,26 @@ mean(A::AbstractArray; dims=:) = _mean(identity, A, dims)
 
 _mean_promote(x::T, y::S) where {T,S} = convert(promote_type(T, S), y)
 
+
+function _promoted_sum(f, A::AbstractArray; dims) # calls f() length(x) +1 times
+    x1 = f(first(A)) / 1
+    result = sum(x -> _mean_promote(x1, f(x)), A; dims)
+end
+function _promoted_sum(f, A::AbstractVector; dims) # calls f() length(x) times
```
Comment (nalimilan) [inline src/Statistics.jl:173]:
```suggestion
function _promoted_sum(f, A::AbstractVector; dims) # calls f(A[1]) twice
```

File: src/Statistics.jl [hunk 5]
```diff
@@ -165,16 +165,26 @@ mean(A::AbstractArray; dims=:) = _mean(identity, A, dims)
 
 _mean_promote(x::T, y::S) where {T,S} = convert(promote_type(T, S), y)
 
+
+function _promoted_sum(f, A::AbstractArray; dims) # calls f() length(x) +1 times
```
Comment (nalimilan) [inline src/Statistics.jl:169]:
```suggestion
function _promoted_sum(f, A::AbstractArray; dims) # calls f(A[1]) twice
```

File: src/Statistics.jl [hunk 6]
```diff
@@ -165,16 +165,26 @@ mean(A::AbstractArray; dims=:) = _mean(identity, A, dims)
...
+    x1 = f(first(A)) / 1
+    result = sum(x -> _mean_promote(x1, f(x)), A; dims)
+end
+function _promoted_sum(f, A::AbstractVector; dims) # calls f() length(x) times
+    x1 = f(first(A)) / 1
+    result = sum(x -> _mean_promote(x1, f(x)), @view A[begin+1:end]; dims,
+                 init = x1)
```
Comment (nalimilan) [inline src/Statistics.jl:176]: This file doesn't use spaces around `=` for keyword arguments. Can you ensure that's the case in your new code?

**Review by nalimilan — COMMENTED** [review]
Thanks!

File: src/Statistics.jl (hunk 1 above)
Comment (kagalenko-m-b) [reply src/Statistics.jl:329]: Do you mean help text only, or the code as well?

File: src/Statistics.jl (hunk 4 above)
Comment (kagalenko-m-b) [reply src/Statistics.jl:173]: That probably should be for line 169, because the method for A::AbstractVector calls  f(A[1]) once

File: src/Statistics.jl (hunk 6 above)
Comment (kagalenko-m-b) [reply src/Statistics.jl:176]: I am also wondering if maybe ```x1 = f(first(A)) / 1``` should be moved out of ```_promoted_sum()```
